### PR TITLE
gradient compression updates

### PR DESCRIPTION
**Pull request by vycezhong**
This PR includes:
1. support PyTorch and Apex.
2. reduce compression overhead for fp16 gradients by converting to fp32 before compressing and converting to fp16 when returned to the application. 
3. improve the random-k algorithm by using the same seed for all workers and servers. Our results show that the improved algorithm can successfully train ResNet-50 on ImageNet without accuracy loss. 
4. achieve workload balance by using the original size as the workload estimation in servers. 
5. add vanilla error-feedback, which does not use learning rate for correction. 
6. add sparse error-feedback for the random-k algorithm. 
7. update docs. 

Bug fixes:
1. fix MXNet's extension linking PyTorch's libraries. (setup.py) 

The PR does not cover passing learning rate to remote servers. It also does not cover hang issue in MXNet. The PR is ready for merge. 

cc: @eric-haibin-lin @szhengac

### CODE REVIEW

File: byteps/common/cpu_reducer.cc [hunk 1]
```diff
@@ -176,262 +139,93 @@ int CpuReducer::sum(void* dst, const void* src1, const void* src2, size_t len,
...
-
-int CpuReducer::_sum_float16(void* dst, const void* src1, const void* src2,
-                             size_t len, float alpha) {
-  // cast src and dst to your float16 type
-  auto in1 = reinterpret_cast<const unsigned short*>(src1);
-  auto in2 = reinterpret_cast<const unsigned short*>(src2);
-  auto out = reinterpret_cast<unsigned short*>(dst);
-  len = len / (size_t)2;
-
-#if __AVX__ && __F16C__
```
Comment (eric-haibin-lin) [inline byteps/common/cpu_reducer.cc:380]: Why removing them?

File: docs/best-practice.md [hunk 2]
```diff
@@ -9,13 +9,13 @@ In non-distributed mode, BytePS is basically doing NCCL allreduce, so it will no
 The only thing you can tune is `BYTEPS_PCIE_SWITCH_SIZE`. If you know your hardware topology, e.g., say you have 8 GPUs in total, 4 GPUs connect to one PCI-e switch, the other 4 GPUs connect to another PCI-e switch, then you should set `BYTEPS_PCIE_SWITCH_SIZE=4`. In this case, you may see 20%~30% performance improvement compared with Horovod/NCCL.
 
 If you have NVLinks, leave `BYTEPS_PCIE_SWITCH_SIZE` unmodified. If you don't know your hardware topology, leave `BYTEPS_PCIE_SWITCH_SIZE` unmodified.
-
+s CPU servers and do not need GPUs. The scheduler can run on any machine.
```
Comment (eric-haibin-lin) [inline docs/best-practice.md:12]: Revert

**Review by eric-haibin-lin — CHANGES_REQUESTED** [review]
This PR is too large. Is it possible to break it down into smaller independent ones?

File: byteps/common/cpu_reducer.cc (hunk 1 above)
Comment (vycezhong) [reply byteps/common/cpu_reducer.cc:380]: We use the version under file.h.
